obspy/io/sac/arrayio.py

```diff
@@ -0,0 +1,637 @@
+                   fmt=native_str("%10d%10d%10d%10d%10d"))
+        for i in range(0, 24, 3):
+            f.write(hs[i:i + 3].data)
+            f.write(b'\n')
+    except:
+        if is_file_name:
+            f.close()
+        raise SacIOError("Cannot write header values: " + f.name)
+
+    if data is not None:
+        npts = hi[9]
+        if npts in (HD.INULL, 0):
+            if is_file_name:
+                f.close()
+            return
+        try:
+            rows = npts // 5
+            np.savetxt(f, np.reshape(data[0:5 * rows], (rows, 5)),
+                       fmt=native_str("%#15.7g%#15.7g%#15.7g%#15.7g%#15.7g"))
+            np.savetxt(f, data[5 * rows:], delimiter=b'\t')
+        except:
+            f.close()
+            raise SacIOError("Cannot write trace values: " + f.name)
+
+    if is_file_name:
+        f.close()
+
+
+# ---------------------- HEADER ARRAY / DICTIONARY CONVERTERS -----------------
+# TODO: this functionality is basically the same as the getters and setters in
+#    sac.sactrace. find a way to avoid duplication?
+# TODO: put these in sac.util?
+def header_arrays_to_dict(hf, hi, hs, nulls=False):
+    """
+    Returns
+    -------
+    dict
+        The correctly-ordered SAC header values, as a dictionary.
+    nulls : bool
+        If True, return all values including nulls.
+
+    """
+    if nulls:
+        items = [(key, val) for (key, val) in zip(HD.FLOATHDRS, hf)] + \
+                [(key, val) for (key, val) in zip(HD.INTHDRS, hi)] + \
+                [(key, val) for (key, val) in zip(HD.STRHDRS, hs)]
+    else:
+        items = [(key, val) for (key, val) in zip(HD.FLOATHDRS, hf) if val != HD.FNULL] + \
+                [(key, val) for (key, val) in zip(HD.INTHDRS, hi) if val != HD.INULL] + \
+                [(key, val) for (key, val) in zip(HD.STRHDRS, hs) if val != HD.SNULL]
+
+    header = dict(items)
+
+    # here, we have to append the 2nd kevnm field into the first and remove
+    #   it from the dictionary.
+    # XXX: kevnm may be null when kevnm2 isn't
+    if 'kevnm2' in header:
+        if 'kevnm' in header:
+            header['kevnm'] += header.pop('kevnm2').decode()
+        else:
+            header['kevnm'] = header.pop('kevnm2').decode()
+
+    return header
+
+
+def dict_to_header_arrays(header=None):
+    """
+    Returns null hf, hi, hs arrays, optionally filled with values from a
+    dictionary.  No header checking.
+
+    """
+    hf, hi, hs = init_header_arrays()
+
+    # have to split kevnm into two fields
+    # TODO: add .lower() to hdr lookups, for safety
+    if header is not None:
+        for hdr, value in header.iteritems():
+            if hdr in HD.FLOATHDRS:
+                hf[HD.FLOATHDRS.index(hdr)] = value
+            elif hdr in HD.INTHDRS:
+                if not isinstance(value, (np.integer, int)):
+                    warnings.warn("Non-integers may be truncated.")
+                    print(" {}: {}".format(hdr, value))
```

Comment:
Right, except warnings only warn once for the same problem.  The user will miss all other key-values that may have been truncated.